### ISSUE TITLE
Java client lib: Refactor to avoid statics

### ISSUE THREAD

**Issue by jerith666** [CarnegieLearningWeb/educational-experiment-service-client-libs#14]
@akpople can chime in and correct me if I'm wrong, but I think that we need to avoid mutable (i.e. non-`final`) `static` fields in order to be able to use this library in a JAX-RS server environment that relies on dependency injection of request-scoped objects.

In particular, the way that `ServiceGenerator` and `Utils.BASE_URL` work is problematic.

I think ideally, we'd have a standard Java factory method that we could invoke like:

```
ExperimentServiceAPI eesApi =
    ExperimentServiceAPIFactory.create("https://upgrade-server.example.com");
```

... and whose return value did not depend on the value of any mutable `static` fields anywhere.

**Comment by jerith666** [comment]
Minor update: a second argument that'll need to be passed in is the token used to authenticate clients.

**Comment by jerith666** [comment]
Another update, based on the discussion in #28: the return type of the proposed factory method would be `ExperimentClient`, not `ExperimentServiceAPI`.

**Comment by dhavalocked** [comment]
There will only be one base URL for client. I think making a factory service for that is redundant. 
I was thinking instead we pass it in constructor and use it in other requests like....

```JAVA
ExperimentClient experimentClient = new ExperimentClient( userId , "authToken" , baseUrl);
experimentClient.setWorkingGroup(workingGroup)
```

If there is a chance that we will have a multiple base_urls for different request then i think we should go with FactoryService methods.
Let me know what you think about it.

**Comment by jerith666** [comment]
You're right that there will only be one base URL for the client.

You're also right that we don't need a factory -- that was appropriate when the thing being returned was an interface (`ExperimentServiceAPI`), but when it's a concrete class, a constructor call like the one you propose works just fine.

**Comment by dhavalocked** [comment]
Fixed in #34

**Comment by jerith666** [comment]
There are some improvements in #34, specifically 1f5a6c4.  However, I'm concerned that the library we're using under the hood here, https://square.github.io/okhttp/, is not really designed to be used in a server-side container.

@akpople, I'm curious to hear your thoughts ... would we be better off building on the JAX-RS client API (https://eclipse-ee4j.github.io/jakartaee-tutorial/jaxrs-client.html)?

**Comment by dhavalocked** [comment]
We are using Retrofit which is a high-level REST abstraction build on top of OkHttp. So yeah all underlying calls would be made using  okhttp. It provides both Sync and Async calls. 

We are using Async with callbacks for our library. is that the reason you think it cant be used in server-side container?

**Comment by jerith666** [comment]
That's part of it, yeah.  There are a couple of things that we need to be concerned about, as I understand it:

### Threading

The JAX-RS container (tomcat) needs to be able to manage all threads that could be started in the course of handling a request.  Since we're making async callbacks, we need to be able to provide hooks that allow the user of this library to control the thread(s) on which those callbacks are executed.

### Caching

OkHttp advertises that it caches responses, but AFAICT it doesn't provide any way to control the lifetimes of those caches.  And it's not clear how long they live, how their eviction policies are managed, etc.  Again, we need the user of this library to be able to manage such things.

### Connection Pooling

OkHttp also says that it uses connection pools and persistent HTTP connections.  Again, a good strategy for improving performance, but we need to make sure the user of the library can control the sizes and lifetimes of these pools.

**Comment by jerith666** [comment]
addressed in #36
